h2o-algos/src/main/java/hex/gam/MatrixFrameUtils/AddGamColumns.java

```diff
@@ -0,0 +1,84 @@
+package hex.gam.MatrixFrameUtils;
+
+import hex.DataInfo;
+import hex.gam.GamSplines.CubicRegressionSplines;
+import water.MRTask;
+import water.MemoryManager;
+import water.fvec.Chunk;
+import water.fvec.Frame;
+import water.fvec.NewChunk;
+
+import static hex.gam.MatrixFrameUtils.GamUtils.locateBin;
+import static hex.gam.MatrixFrameUtils.GenerateGamMatrixOneColumn.updateAFunc;
+import static hex.gam.MatrixFrameUtils.GenerateGamMatrixOneColumn.updateFMatrixCFunc;
+
+/**
+ * Given a Frame, the class will add all the gam columns to the end of the Frame right before
+ * the response column if it exists
+ */
+public class AddGamColumns extends MRTask<AddGamColumns> {
+  double[][][] _binvD;
```

Comment:
This class build the model matrix X that contains the predictors without basis function expansion and basis function expansion for all columns specified in _parms._gam_X.  It performs the operation in 6.1.  However, the matrices binvD and Z matrices if needed are already computed.